lib/ansible/modules/cloud/vmware/vcenter_folder.py

```diff
@@ -176,49 +176,30 @@ def ensure(self):
         folder_name = self.module.params.get('folder_name')
         parent_folder = self.module.params.get('parent_folder', None)
         results = {'changed': False, 'result': {}}
-        if state == 'present':
-            # Check if the folder already exists
-            p_folder_obj = None
-            if parent_folder:
-                if "/" in parent_folder:
-                    parent_folder_parts = parent_folder.strip('/').split('/')
-                    p_folder_obj = None
-                    for part in parent_folder_parts:
-                        part_folder_obj = self.get_folder(datacenter_name=datacenter_name,
-                                                          folder_name=part,
-                                                          folder_type=folder_type,
-                                                          parent_folder=p_folder_obj)
-                        if not part_folder_obj:
-                            self.module.fail_json(msg="Could not find folder %s" % part)
-                        p_folder_obj = part_folder_obj
-
-                else:
-                    p_folder_obj = self.get_folder(datacenter_name=datacenter_name,
-                                                   folder_name=parent_folder,
-                                                   folder_type=folder_type)
 
-                    if not p_folder_obj:
-                        self.module.fail_json(msg="Parent folder %s does not exist" % parent_folder)
-
-                    # Check if folder exists under parent folder
-                    child_folder_obj = self.get_folder(datacenter_name=datacenter_name,
-                                                       folder_name=folder_name,
-                                                       folder_type=folder_type,
-                                                       parent_folder=p_folder_obj)
-                    if child_folder_obj:
-                        results['result']['path'] = self.get_folder_path(child_folder_obj)
-                        results['result'] = "Folder %s already exists under" \
-                                            " parent folder %s" % (folder_name, parent_folder)
-                        self.module.exit_json(**results)
-            else:
-                folder_obj = self.get_folder(datacenter_name=datacenter_name,
-                                             folder_name=folder_name,
-                                             folder_type=folder_type)
-
-                if folder_obj:
-                    results['result']['path'] = self.get_folder_path(folder_obj)
-                    results['result']['msg'] = "Folder %s already exists" % folder_name
-                    self.module.exit_json(**results)
+        # None if parent_folder is None else parent folder object
+        p_folder_obj = self.get_parent_folder_obj(datacenter_name=datacenter_name,
```

Comment:
Both branches declared p_folder_obj as None and set parent object to it if parent folder was defined and found. I combined this logic to on function to reduce complexity.